File: scala-3/tax-system/src/main/scala/com/rbleggi/taxsystem/specification/DefaultTaxSpecification.scala
package com.rbleggi.taxsystem.specification

import com.rbleggi.taxsystem.model.{Product, TaxConfiguration}

class DefaultTaxSpecification(config: TaxConfiguration) extends TaxSpecification {
  override def isSatisfiedBy(state: String, year: Int): Boolean =
    config.state == state && config.year == year

  override def calculateTax(product: Product, price: Double): Double =
    price * config.rates.getOrElse(product.category, 0.10)
}

File: java/tax-system/src/main/java/com/rbleggi/taxsystem/specification/DefaultTaxSpecification.java
package com.rbleggi.taxsystem.specification;

import com.rbleggi.taxsystem.model.Product;
import com.rbleggi.taxsystem.model.TaxConfiguration;

public class DefaultTaxSpecification implements TaxSpecification {
    private final TaxConfiguration config;

    public DefaultTaxSpecification(TaxConfiguration config) {
        this.config = config;
    }

    @Override
    public boolean isSatisfiedBy(String state, int year) {
        return config.state().equals(state) && config.year() == year;
    }

    @Override
    public double calculateTax(Product product, double price) {
        return price * config.rates().getOrDefault(product.category(), 0.0);
    }
}
